### ISSUE TITLE
TSAN error when parsing URLs

### ISSUE THREAD

**Issue by shadowfacts** [karwa/swift-url#166]
I was running with the thread sanitizer enabled to test something else, and ran into a crash in WebURL's parsing code. It happens at this line: https://github.com/karwa/swift-url/blob/main/Sources/WebURL/Parser/Parser.swift#L74  because the closures passed in as the left/right params to map both try to capture mutable references to `callback` with the ampersand operator.

Even though left and right are never both called, this is still an exclusivity violation:
> TSAN is trying to be helpful by checking for simultaneous inout operations on the property, _even if we never perform an actual memory access_. Simply accessing the property with & triggers the race. What's more it is technically Swift's right to crash with an exclusivity error, though it doesn't happen to do so as of this writing. 
http://www.russbishop.net/the-law

The solution, I think, is to just lift the switch in `Either.map` out into `_urlFromBytes_impl` so that the two closures don't need to be created.

**Comment by shadowfacts** [comment]
I'm using the Xcode 14 RC.

Oddly, I'm not longer able to reproduce the outright crash in my app and I'm getting (almost) the same runtime warning you're seeing. When I reported this yesterday, the crash was 100% repro-able (and hand-inling the map function fixed it).

This is the warning I'm seeing (and it only reports an access from a single thread):
![image](https://user-images.githubusercontent.com/7091588/189727155-2636b0d8-777c-4104-8907-294675d237b1.png)

But, using this inside of an XCTestCase in a Swift package that depends on WebURL (being run from an xcodeproj that includes the package):
```swift
func testWebURL() {
    DispatchQueue.concurrentPerform(iterations: 100_000) { i in
        _ = WebURL("https://example.com/foo/\(i)")
    }
}
```
I am able to reproduce the crash consistently when running the test with the thread sanitizer enabled in the Test scheme configuration.
![image](https://user-images.githubusercontent.com/7091588/189728659-adb1e3d1-b141-4457-be27-5be4b2aa4ee4.png)

Before the crash, I sometimes (not consistently) get the same runtime sanitizer warning and two threads are reported accessing the same variable (although the report printing in the console says the variable being access is at `0x780000000000` which seems like a suspiciously round number).

Using the same code, but in the main function of a standalone Swift package, I'm not able to reproduce either the warning or the crash when running with `--sanitize=thread`.

**Comment by karwa** [comment]
OK, so this very much looks like a bug in TSan. 

I have a branch with a fix, `tsan-workaround` (PR #168). Would you be able to try it and confirm that it fixes the issue for you?

**Comment by shadowfacts** [comment]
That branch does indeed fix the issue.

**Comment by karwa** [comment]
Should be fixed in 0.4.1. Feel free to reopen if it is not fixed.
